### PR TITLE
Stop retry when err is a clientInteraction

### PR DESCRIPTION
**Pull request by duanxianbo**
(task:13603)

### CODE REVIEW

File: src/libs/retry/infinite.test.ts [hunk 1]
```diff
@@ -6,85 +6,133 @@ import { ErrorCode } from '@lib/error/errors';
...
+    CLIENT_INTERACTION_ERR,
+    LONG_DELAY,
+    OUTAGE_ERR,
+    RANDOM_OFFSET,
+    SHORT_DELAY,
+    TRANSIENT_ERR,
+} from './testing';
 
-describe('InfiniteRetry', () => {
+describe.only('InfiniteRetry', () => {
```
Comment (naturetime10) [inline src/libs/retry/infinite.test.ts]:
```suggestion
describe('InfiniteRetry', () => {
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: src/libs/retry/infinite.ts [hunk 2]
```diff
@@ -1,39 +1,64 @@
...
-        private backoff: BackOff,
+        private logger: Logger,
+        private shortBackOff: BackOff,
+        private longBackOff: BackOff,
         private beforeSkipRetry?: () => Promise<void>,
     ) {}
 
     public withRetry(
         execute: () => Promise<InternalError | undefined>,
+        context?: Record<string, string>,
```
Comment (naturetime10) [inline src/libs/retry/infinite.ts]:
```suggestion
        context: Record<string, string>,
        execute: () => Promise<InternalError | undefined>,
```

File: src/libs/retry/infinite.ts [hunk 3]
```diff
@@ -1,39 +1,64 @@
...
-                if (errorCategory === 'ClientInteraction') {
-                    await this.beforeSkipRetry?.call(null);
-                    resolve([retries, err]);
-                    return;
-                }
+                switch (errorCategory) {
+                    case 'ClientInteraction':
+                        await this.beforeSkipRetry?.call(null);
+                        resolve([retries, err]);
+
```
Comment (naturetime10) [inline src/libs/retry/infinite.ts]:
```suggestion
```

File: src/libs/retry/testing.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+
```
Comment (naturetime10) [inline src/libs/retry/testing.ts]:
```suggestion
```

File: src/libs/retry/retry.ts [hunk 5]
```diff
@@ -1,7 +1,7 @@
 import { InternalError } from '@lib/error/errors';
-
 export interface Retry {
     withRetry: (
         execute: () => Promise<InternalError | undefined>,
+        context?: Record<string, string>,
```
Comment (naturetime10) [inline src/libs/retry/retry.ts]:
```suggestion
        context: Record<string, string>,
        execute: () => Promise<InternalError | undefined>,
```

File: src/libs/retry/maxCount.ts [hunk 6]
```diff
@@ -1,42 +1,69 @@
...
+                            this.shortBackOff.delay(),
+                        );
+
+                        break;
+                    case 'Outage':
+                        this.longBackOff.onFailure();
+                        await this.runtime.threadSleep(
+                            this.longBackOff.delay(),
+                        );
+
```
Comment (naturetime10) [inline src/libs/retry/maxCount.ts]:
```suggestion
```

File: src/libs/retry/maxCount.ts [hunk 7]
```diff
@@ -1,42 +1,69 @@
...
+                        await this.beforeSkipRetry?.call(null);
+                        resolve([retries, err]);
+
+                        return;
+                    case 'Transient':
+                        this.shortBackOff.onFailure();
+                        await this.runtime.threadSleep(
+                            this.shortBackOff.delay(),
+                        );
+
```
Comment (naturetime10) [inline src/libs/retry/maxCount.ts]:
```suggestion
```

File: src/libs/retry/maxCount.ts [hunk 8]
```diff
@@ -1,42 +1,69 @@
...
+                    [ErrCategoryProp]: errorCategory,
+                }, context);
 
-                this.backoff.onFailure();
-                await this.runtime.threadSleep(this.backoff.delay());
+                switch (errorCategory) {
+                    case 'ClientInteraction':
+                        await this.beforeSkipRetry?.call(null);
+                        resolve([retries, err]);
+
```
Comment (naturetime10) [inline src/libs/retry/maxCount.ts]:
```suggestion
```

File: src/libs/retry/maxCount.ts [hunk 9]
```diff
@@ -1,42 +1,69 @@
...
+        private logger: Logger,
+        private shortBackOff: BackOff,
+        private longBackOff: BackOff,
         private maxCount: number,
         private beforeSkipRetry?: () => Promise<void>,
     ) {}
 
     public withRetry(
         execute: () => Promise<InternalError | undefined>,
+        context?: Record<string, string>,
```
Comment (naturetime10) [inline src/libs/retry/maxCount.ts]:
```suggestion
        context: Record<string, string>,
        execute: () => Promise<InternalError | undefined>,
```

File: src/libs/retry/timeout.ts [hunk 10]
```diff
@@ -1,22 +1,30 @@
...
+        private shortBackOff: BackOff,
+        private longBackOff: BackOff,
         private duration: number,
         private beforeRetryDelay?: () => Promise<void>,
         private beforeSkipRetry?: () => Promise<void>,
     ) {}
 
     public withRetry(
         execute: () => Promise<InternalError | undefined>,
+        context?: Record<string, string>,
```
Comment (naturetime10) [inline src/libs/retry/timeout.ts]:
```suggestion
        context: Record<string, string>,
        execute: () => Promise<InternalError | undefined>,
```

File: src/libs/retry/timeout.ts [hunk 11]
```diff
@@ -25,23 +33,41 @@ export class TimeoutRetry implements Retry {
...
+                    [BodyProp]: err.message,
+                }, context)
+                await this.beforeRetryDelay?.call(null);
+                const errorCategory = getErrCategory(err);
+                this.logger.logWithContext('Warning', {
+                    [ErrCategoryProp]: errorCategory,
+                }, context);
+
+                let duration;
+
```
Comment (naturetime10) [inline src/libs/retry/timeout.ts]:
```suggestion
```

File: src/libs/retry/timeout.ts [hunk 12]
```diff
@@ -25,23 +33,41 @@ export class TimeoutRetry implements Retry {
...
-                    return;
+                this.logger.logWithContext('Error', {
+                    [ErrCodeProp]: err.errorCode,
+                    [BodyProp]: err.message,
+                }, context)
+                await this.beforeRetryDelay?.call(null);
+                const errorCategory = getErrCategory(err);
+                this.logger.logWithContext('Warning', {
+                    [ErrCategoryProp]: errorCategory,
+                }, context);
```
Comment (naturetime10) [inline src/libs/retry/timeout.ts]:
```suggestion
                const errorCategory = getErrCategory(err);
                this.logger.logWithContext('Warning', {
                    [ErrCategoryProp]: errorCategory,
                }, context);
                await this.beforeRetryDelay?.call(null);
```

File: src/libs/telemetry/Props.ts [hunk 13]
```diff
@@ -14,5 +14,7 @@ export const PathProp = 'Path';
 export const ProtocolProp = 'Protocol';
 export const CommitProp = 'Commit';
 export const HostProp = 'Host';
+export const ErrCodeProp = 'ErrCode';
```
Comment (naturetime10) [inline src/libs/telemetry/Props.ts]:
```suggestion
export const ErrCodeProp = 'ErrorCode';
```

File: src/libs/telemetry/Props.ts [hunk 14]
```diff
@@ -14,5 +14,7 @@ export const PathProp = 'Path';
...
+export const ErrCategoryProp = 'ErrCategory';
```
Comment (naturetime10) [inline src/libs/telemetry/Props.ts]:
```suggestion
export const ErrCategoryProp = 'ErrorCategory';
```

File: src/libs/telemetry/TelemetryClient.ts [hunk 15]
```diff
@@ -4,7 +4,12 @@ import { Optional } from '@lib/sugar/optional';
 
 import { Props } from './Props';
 
-export class TelemetryClient {
+
+export interface TelemetryClientInterface {
+    uploadLog(logEntries: Props[]): Promise<[Optional<HTTPResponse>, Optional<InternalError>]>;
+}
+
```
Comment (naturetime10) [inline src/libs/telemetry/TelemetryClient.ts]:
```suggestion
```

File: src/libs/telemetry/TelemetryClient.ts [hunk 16]
```diff
@@ -4,7 +4,12 @@ import { Optional } from '@lib/sugar/optional';
...
+export class TelemetryClient implements TelemetryClientInterface {
```
Comment (naturetime10) [inline src/libs/telemetry/TelemetryClient.ts]:
```suggestion
export class TelemetryClient {
```

File: src/libs/telemetry/TelemetryClient.ts [hunk 17]
```diff
@@ -4,7 +4,12 @@ import { Optional } from '@lib/sugar/optional';
 
 import { Props } from './Props';
 
-export class TelemetryClient {
+
```
Comment (naturetime10) [inline src/libs/telemetry/TelemetryClient.ts]:
```suggestion
```

File: src/libs/telemetry/TelemetryClient.ts [hunk 18]
```diff
@@ -24,3 +29,9 @@ export class TelemetryClient {
         );
     }
 }
+
+export class TestTelemetryClient implements TelemetryClientInterface{
+    public uploadLog(logEntries: Props[]): Promise<[Optional<HTTPResponse>, Optional<InternalError>]> {
+        return Promise.resolve([undefined, undefined]);
+    }
+}
```
Comment (naturetime10) [inline src/libs/telemetry/TelemetryClient.ts]:
```suggestion
```

I think we can create and pass in `fakeHttpClient` instead.

File: src/libs/telemetry/RawLogger.ts [hunk 19]
```diff
@@ -4,7 +4,7 @@ import { Runtime } from '@lib/runtime/runtime';
 import { LOG_LEVEL_RANKS, LogLevel } from './LogLevel';
 import { Logger } from './Logger';
 import { HappenAtProp, Props, SeverityProp } from './Props';
-import { TelemetryClient } from './TelemetryClient';
+import { TelemetryClient, TelemetryClientInterface } from './TelemetryClient';
```
Comment (naturetime10) [inline src/libs/telemetry/RawLogger.ts]:
```suggestion
import { TelemetryClient } from './TelemetryClient';
```

File: src/libs/telemetry/RawLogger.ts [hunk 20]
```diff
@@ -14,7 +14,7 @@ export class RawLogger implements Logger {
     constructor(
         private visibleLevel: LogLevel,
         private runtime: Runtime,
-        telemetryClient: TelemetryClient,
+        telemetryClient: TelemetryClientInterface,
```
Comment (naturetime10) [inline src/libs/telemetry/RawLogger.ts]:
```suggestion
        telemetryClient: TelemetryClient,
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

**Comment by naturetime10** [comment]
More comments will be provided for test code.

File: src/libs/retry/infinite.test.ts [hunk 21]
```diff
@@ -6,107 +6,149 @@ import { ErrorCode } from '@lib/error/errors';
...
     let beforeSkipRetry: () => Promise<void>;
 
     beforeEach(() => {
         testRuntime = new TestRuntime();
         beforeSleepSemaphore = new Semaphore('BeforeSleep');
-        beforeSkipRetrySemaphore = new Semaphore('BeforeSkipRetry');
         exitSemaphore = new Semaphore('Exit');
         beforeSkipRetry = async () => {
-            beforeSkipRetrySemaphore.release();
+            beforeSleepSemaphore.release();
```
Comment (naturetime10) [inline src/libs/retry/infinite.test.ts]: Is calling `beforeSleepSemaphore` expected?

File: src/libs/retry/infinite.test.ts [hunk 22]
```diff
@@ -6,107 +6,149 @@ import { ErrorCode } from '@lib/error/errors';
...
+                durations: [
+                    SHORT_DELAY * 2 + RANDOM_OFFSET,
+                    LONG_DELAY * 2 + RANDOM_OFFSET,
+                    LONG_DELAY * 2 + RANDOM_OFFSET,
+                ],
+                expectRetries: 3,
+                expectErr: {
+                    errorCode: CLIENT_INTERACTION_ERR,
+                },
+                sleepAwakeCount: 2,
```
Comment (naturetime10) [inline src/libs/retry/infinite.test.ts]:
```suggestion
                sleepAwakeCount: 2,
                expectRetries: 3,
                expectErr: {
                    errorCode: CLIENT_INTERACTION_ERR,
                },
```

File: src/libs/retry/infinite.test.ts [hunk 23]
```diff
@@ -6,107 +6,149 @@ import { ErrorCode } from '@lib/error/errors';
...
+                durations: [
+                    SHORT_DELAY * 2 + RANDOM_OFFSET,
+                    LONG_DELAY * 2 + RANDOM_OFFSET,
+                    SHORT_DELAY * 4 + RANDOM_OFFSET,
+                    LONG_DELAY * 4 + RANDOM_OFFSET,
+                    LONG_DELAY * 4 + RANDOM_OFFSET,
+                ],
+                expectRetries: 5,
+                expectErr: undefined,
+                sleepAwakeCount: 4,
```
Comment (naturetime10) [inline src/libs/retry/infinite.test.ts]:
```suggestion
                sleepAwakeCount: 4,
                expectRetries: 5,
                expectErr: undefined,
```

File: src/libs/retry/infinite.test.ts [hunk 24]
```diff
@@ -6,107 +6,149 @@ import { ErrorCode } from '@lib/error/errors';
...
-                const [retriesRes, errorRes] = await infinite.withRetry(
+                const [retriesRes, errorRes] = await infinite.withRetry({},
                     executable.execute.bind(executable),
                 );
 
-                strictEqual(retriesRes, retries);
-                strictEqual(duration.totalMilliSeconds, 600);
-                strictEqual(errorRes?.errorCode, expectedErrorCode);
+                strictEqual(retriesRes, expectRetries);
+                strictEqual(errorRes?.errorCode, expectErr?.errorCode);
```
Comment (naturetime10) [inline src/libs/retry/infinite.test.ts]:
```suggestion
                strictEqual(expectRetries, retriesRes);
                strictEqual(expectErr?.errorCode, errorRes?.errorCode);
```

File: src/libs/retry/maxCount.test.ts [hunk 25]
```diff
@@ -6,130 +6,167 @@ import { ErrorCode } from '@lib/error/errors';
...
+                minDelay: SHORT_DELAY,
+                randomOffset: 1,
             });
-            const maxCount = new MaxCountRetry(
+            const longBackOff = new ExponentialBackoff(testRandGen, {
+                minDelay: LONG_DELAY,
+                randomOffset: 1,
+                resetOnSuccess: true,
+            });
+
```
Comment (naturetime10) [inline src/libs/retry/maxCount.test.ts]:
```suggestion
```

File: src/libs/retry/timeout.test.ts [hunk 26]
```diff
@@ -47,106 +53,144 @@ describe('Timeout', () => {
...
+                resetOnSuccess: true,
             });
 
-            const timeout = new TimeoutRetry(
+            beforeRetryDelay = async () => {
+                testClock.setCurrentTime(
+                    new Date(
+                        testClock.now().getTime() +
+                            executeDuration[executable.count - 1],
+                    ),
```
Comment (naturetime10) [inline src/libs/retry/timeout.test.ts]: Can we put this into a new variable called `newDate`

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: src/libs/retry/infinite.test.ts (hunk 21 above)
Comment (duanxianbo) [reply src/libs/retry/infinite.test.ts]: it is used for testing purpose, we have a loop to check the runtime sleep and awake behaviours, this is used to trigger before sleep and after sleep signals, for client interaction err, we donot call runtime.sleep(no need to sleep) to release this semaphore, instead we use this beforeSkipRetry to release it

File: src/libs/network/middleware/clientHTTPWithRetry.ts [hunk 27]
```diff
@@ -16,6 +16,7 @@ export function clientHTTPWithRetry(makeRetry: ()=>Retry): Middleware<HTTPClient
                 let result: [Optional<HTTPResponse>, Optional<InternalError>];
                 const retry = makeRetry();
                 await retry.withRetry(
+                    context!,
```
Comment (naturetime10) [inline src/libs/network/middleware/clientHTTPWithRetry.ts]:
```suggestion
                    context || {},
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

**Review by naturetime10 — APPROVED** [review]

**Pull request merged**